### PR TITLE
Add `dhall to-directory-tree` support for `Map`s

### PR DESCRIPTION
**Pull request by Gabriel439**
Fixes https://github.com/dhall-lang/dhall-haskell/issues/1704

### CODE REVIEW

**Comment by Gabriel439** [comment]
@sjakobi: Personally, I agree with what @Profpatsch suggested here:

https://github.com/dhall-lang/dhall-haskell/issues/1633#issuecomment-575762860

... which is that we should reject keys with `/` in their name and just provide an error message explaining how to restructure things.  For the issue that @timbertson mentioned in:

https://github.com/dhall-lang/dhall-haskell/issues/1633#issuecomment-575875577

... that can be resolved through using the newly-added `with` keyword as the example code would now become:

```dhall
let dockerStuff =
      let base = ./Docker.dhall

      in  base with script.`optionalStuff.sh` = None Text

in  …
```

**Comment by bfrk** [comment]
Wow, that was fast! Thanks!

**Comment by Gabriel439** [comment]
@bfrk: You're welcome! 🙂

File: dhall/src/Dhall/DirectoryTree.hs [hunk 1]
```diff
@@ -99,9 +105,29 @@ toDirectoryTree path expression = case expression of
...
+
+    extract (RecordLit [("mapKey", TextLit (Chunks [] key)), ("mapValue", value)]:records) = do
+        fmap ((key, value) :) (extract records)
+
+    extract _ = do
+        empty
 
-        Exception.throwIO FilesystemError{..}
+    process key value = do
+        if Text.isInfixOf "/" key
```
Comment (sjakobi) [inline dhall/src/Dhall/DirectoryTree.hs]: How about also disallowing `\` on Windows?

**Comment by bfrk** [comment]
I just noted that the documentation (or rather: the long error message) doesn't yet mention the possibility to use a dynamic map.

**Comment by Gabriel439** [comment]
@sjakobi: The main reason I'm going with the record route for now is to take advantage of Dhall's support for record operations (e.g. `with`, `∧`, etc.).  However, I could imagine that at some point we do something similar to the JSON/YAML conversions which is provide a type analogous to `Prelude.JSON.Type` to handle the weakly-typed general case, while still providing support for records in the more common case.

**Comment by timbertson** [comment]
> I agree that disallowing path separators in keys is the more principled approach!

But.. doesn't that mean it's impossible to create nested directories when using a map?

i.e. with a record you can do `{ foo.bar = "baz" }`, to create `foo/bar`. But it looks like it's impossible to do that with a Map. That doesn't feel right.

To me, the better solution would be to have the shapes be different for records / maps. So this as a record:

```
{
  foo = { bar = "contents" },
  other = "more contents"
}
```

Could also be expressed as:

```
toMap {
  `foo/bar` = "contents",
  other = "more contents"
}
```

You'd need to validate the whole map to ensure you didn't have an implicit path which is also a key in the map (i.e. conflicting directory/file), but that seems like the right thing to do, instead of disallowing nested files.

**Comment by timbertson** [comment]
One awkward thing would be that there's no way to convert between the two forms (i.e. flatten ested records into a map). You _could_ add `toFlatMap` to the language which takes a `\(path: List Text)` in order to determine what the flattened key for a nested value should be, but that seems pretty niche.

**Comment by Gabriel439** [comment]
@timbertson: You can still create nested directories using nested `Map`s, too, like this:

```dhall
toMap { foo = toMap { bar = "baz" } }
```

**Comment by timbertson** [comment]
But only if all of your records are at the same nesting level, you can't have one file in a subdirectory and another file at the toplevel:

```
toMap {
  foo = toMap { bar = "contents" },
  other = "more contents"
}
```

> Error: ❰toMap❱ expects a homogenous record

**Comment by Gabriel439** [comment]
@timbertson: So I think the answer to that is the solution @sjakobi suggested in https://github.com/dhall-lang/dhall-haskell/pull/1705#pullrequestreview-374543434 (analogous to how `Prelude.JSON.Type` works), but that's outside of the scope of this pull request

**Comment by timbertson** [comment]
If @sjakobi's suggestion (which seems good to me) is later implemented, would there then be three different types of structure supported by `to-directory-tree`? Or would you drop the Map version? It just doesn't seem like this Map version is worth the complexity/confusion if it can only produce flat (or uniformly-deep) file structures. @bfrk out of curiosity, is your use case only for flat maps?

As an aside: I think it'd make it easier to explain / document if this _only_ supported `Map Text Text`, and not arbitrarily deep `Map Text (Map Text Text)` etc

**Comment by Gabriel439** [comment]
@timbertson: We'd keep all three of records, `Map`s, and the weakly-typed `DirectoryTree` structure.  Think of this tool as analogous to the JSON/YAML conversion utilities, which also do the same (except with `Prelude.JSON.Type` instead of `DirectoryTree`)

**Comment by bfrk** [comment]
@timbertson:
> @bfrk out of curiosity, is your use case only for flat maps?

The directory structure in my case is indeed flat: one directory per server machine and under that one file per service running on that machine. However, I have also been thinking about using a directory structure that is one level deeper, where each service gets its own directory and the files in there correspond to individual configuration options. This would be useful to support completely parser-less service configuration a la https://cr.yp.to/daemontools/envdir.html (meaning: the service itself, usually written in C, doesn't have to parse text at all). To support both styles in parallel, homogenous `Map`s would indeed be insufficient, and I would need an explicit `DirectoryTree` structure.

**Comment by timbertson** [comment]
> We'd keep all three of records, Maps, and the weakly-typed DirectoryTree structure.

I guess that's OK, if you can mix them you can do e.g. `{ some = { directory = toMap { ... } } }`. And now I understand why you don't want to have different rules for record / map keys -- I had assumed if you were using a map, the whole structure had to be just a map.

If you did require the whole structure to be a `Map Text Text` (instead of allowing maps inside records), would it then make sense for this "map mode" to support `/` in keys as a path separator? I'd prefer that tradeoff personally.

It just feels very jarring with the current implementation that you suddenly _can't_ use a map as soon as any map entry has a sibling directory.

**Pull request merged**